File: second/Moves.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.Deque;

class Node {
    int[] position;
    int value;
    List<Node> children;

    Node(int[] position, int value) {
        this.position = position;
        this.value = value;
        this.children = new ArrayList<>();
    }
}

public class Moves {
    public static void printTree(Node node) {
        if (node == null) return;

        System.out.println("Position: " + node.position[0] + ", " + node.position[1] + ", Value: " + node.value);

        for (Node child : node.children) {
            printTree(child);
        }
    }

    public static int[][] findNeighbors(int[] position, int N) {
        int i = position[0];
        int j = position[1];

        int[][] neighborPositions = {
                {-1, -1}, {-1, 0}, {-1, 1},
                {0, -1},          {0, 1},
                {1, -1}, {1, 0}, {1, 1}
        };

        List<int[]> neighbors = new ArrayList<>();

        for (int[] neighbor : neighborPositions) {
            int new_r = i + neighbor[0];
            int new_c = j + neighbor[1];

            if (new_r >= 0 && new_r < N && new_c >= 0 && new_c < N) {
                neighbors.add(new int[]{new_r, new_c});
            }
        }

        return neighbors.toArray(new int[0][0]);
    }

    public static List<Node> keepPossible(int[][] neighbors, int value, int[][] matrix, Set<String> visited) {
        List<Node> possibleNeighbors = new ArrayList<>();

        for (int[] n : neighbors) {
            int i = n[0];
            int j = n[1];
            if (value > matrix[i][j] && !visited.contains(i + "," + j)) {
                Node node = new Node(n, matrix[i][j]);
                possibleNeighbors.add(node);
            }
        }

        return possibleNeighbors;
    }

    public static void expandTree(Node currentNode, int N, int[][] matrix, Set<String> visited) {
        visited.add(currentNode.position[0] + "," + currentNode.position[1]);
        int[][] neighbors = findNeighbors(currentNode.position, N);
        List<Node> possibleNeighbors = keepPossible(neighbors, currentNode.value, matrix, visited);
        currentNode.children.addAll(possibleNeighbors);
        for (Node child : currentNode.children) {
            expandTree(child, N, matrix, new HashSet<>(visited));
        }
    }

    public static List<int[]> bfs(Node root, int N) {
        Deque<NodePathPair> queue = new ArrayDeque<>();
        queue.addLast(new NodePathPair(root, new ArrayList<>() {{
            add(root.position);
        }}));

        while (!queue.isEmpty()) {
            NodePathPair pair = queue.removeFirst();
            Node currentNode = pair.node;
            List<int[]> path = pair.path;

            if (currentNode.position[0] == N - 1 && currentNode.position[1] == N - 1) {
                return path;
            }

            for (Node child : currentNode.children) {
                List<int[]> newPath = new ArrayList<>(path);
                newPath.add(child.position);
                queue.addLast(new NodePathPair(child, newPath));
            }
        }

        return null;
    }

    static class NodePathPair {
        Node node;
        List<int[]> path;

        NodePathPair(Node node, List<int[]> path) {
            this.node = node;
            this.path = path;
        }
    }

    public static String pathToMoves(List<int[]> path) {
        if (path == null) return null;
        else {
            StringBuilder directions = new StringBuilder("[");
            for (int i = 1; i < path.size(); i++) {
                int[] currentStep = path.get(i);
                int[] prevStep = path.get(i - 1);
                int dr = currentStep[0] - prevStep[0];
                int dc = currentStep[1] - prevStep[1];
                String direction = "";
                if (dr == -1) direction += "N";
                else if (dr == 1) direction += "S";
                if (dc == 1) direction += "E";
                else if (dc == -1) direction += "W";
                directions.append(direction).append(",");
            }
            directions.deleteCharAt(directions.length() - 1);
            directions.append("]");
            return directions.toString();
        }
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Wrong usage.");
            System.exit(1);
        }

        String filename = args[0];
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            int N = Integer.parseInt(br.readLine());
            int[][] matrix = new int[N][N];

            for (int i = 0; i < N; i++) {
                String[] values = br.readLine().split(" ");
                for (int j = 0; j < N; j++) {
                    matrix[i][j] = Integer.parseInt(values[j]);
                }
            }

            Node root = new Node(new int[]{0, 0}, matrix[0][0]);
            Set<String> visited = new HashSet<>();
            visited.add("0,0");

            expandTree(root, N, matrix, visited);

            List<int[]> path = bfs(root, N);

            String result = pathToMoves(path);

            if (result != null) {
                System.out.println(result);
            } else {
                System.out.println("IMPOSSIBLE");
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: second/moves.pl
:- dynamic(node/2).

% Entry point
moves(FileName, Moves) :-
    read_input(FileName, N, Matrix),
    assert_initial_node(N, Matrix),
    nth0(0, Matrix, Row),
    nth0(0, Row, CurrentValue),
    X is N - 1,
    bfs([[node((0, 0), CurrentValue)]], N, Matrix, node((X, X), _), [], Path),
    (Path = [] -> false ; path_to_moves(Path, Moves)).

% Read the input file
read_input(FileName, N, Matrix) :-
    open(FileName, read, Stream),
    read_line_to_codes(Stream, NLine),
    atom_codes(AtomN, NLine),
    atom_number(AtomN, N),
    read_matrix(Stream, N, Matrix),
    close(Stream).

read_matrix(_, 0, []).
read_matrix(Stream, N, [Row | Matrix]) :-
    N > 0,
    read_line_to_codes(Stream, Line),
    atom_codes(AtomLine, Line),
    atomic_list_concat(Atoms, ' ', AtomLine),
    maplist(atom_number, Atoms, Row),
    Next is N - 1,
    read_matrix(Stream, Next, Matrix).

% Initialize the root node
assert_initial_node(_, Matrix) :-
    nth0(0, Matrix, [Value | _]),
    assert(node((0, 0), Value)).

% Breadth-first search

bfs([], _, _, _, _, _) :- !.

bfs([CurrentPath | RestPaths], N, Matrix, Goal, Visited, Path) :-
    CurrentPath = [node((I, J), _) | _],
    append(Visited, [(I,J)], NewVisited),
    (
        (I is N-1, J is N-1) ->
            Path = CurrentPath
        ;
            nth0(I, Matrix, Row),
            nth0(J, Row, Value),
            find_neighbors((I, J), N, Neighbors),
            keep_possible(Neighbors, Value, Matrix, NewVisited, PossibleNeighbors),
            expand_paths(CurrentPath, PossibleNeighbors, NewPaths),
            append(RestPaths, NewPaths, UpdatedPaths),
            bfs(UpdatedPaths, N, Matrix, Goal, NewVisited, Path)
    ).


expand_paths(_, [], []).
expand_paths(Path, [Neighbor | Neighbors], [[Neighbor | Path] | Expanded]) :-
    expand_paths(Path, Neighbors, Expanded).

% Find neighbors within the grid
find_neighbors((I, J), N, Neighbors) :-
    neighbor_position(NeighborPositions),
    findall((NewI, NewJ), (
        member((DI, DJ), NeighborPositions),
        NewI is I + DI,
        NewJ is J + DJ,
        NewI >= 0, NewI < N,
        NewJ >= 0, NewJ < N
    ), Neighbors).

% Neighbor positions (8 directions)
neighbor_position([(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]).

% Keep possible neighbors with values less than the current node
keep_possible([], _, _, _, []).
keep_possible([(I, J) | Neighbors], Value, Matrix, Visited, [node((I, J), NeighborValue) | Result]) :-
    nth0(I, Matrix, Row),
    nth0(J, Row, NeighborValue),
    NeighborValue < Value,
    \+ member((I, J), Visited),
    keep_possible(Neighbors, Value, Matrix, Visited, Result).
keep_possible([_ | Neighbors], Value, Matrix, Visited, Result) :-
    keep_possible(Neighbors, Value, Matrix, Visited, Result).

% Convert path to moves
path_to_moves(Path, Moves) :-
    reverse(Path, ReversedPath),
    revpath_to_moves(ReversedPath, Moves).

% Translate moves to direction symbols
revpath_to_moves([_], []).
revpath_to_moves([node((I1, J1), _), node((I2, J2), _) | T], [Direction | Rest]) :-
    direction_symbol(I1, J1, I2, J2, Direction),
    revpath_to_moves([node((I2, J2), _) | T], Rest).

% Map coordinates to direction symbols
direction_symbol(I1, J1, I2, J2, Symbol) :-
    DX is I2 - I1,
    DY is J2 - J1,
    direction(DX, DY, Symbol).

% Direction predicates
direction(1, 1, se).
direction(0, 1, e).
direction(1, 0, s).
direction(-1, -1, nw).
direction(0, -1, w).
direction(-1, 0, n).
direction(-1, 1, ne).
direction(1, -1, sw).